### PR TITLE
Refactoring code for application printing + fixing #11091 (inconsistencies in parsing/printing notations to partial applications)

### LINKED ISSUE #4690: Implicit arguments are inconsistently parsed in notations
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#4690
From: @maximedenes
Reported version: 8.5
CC:   @ejgallego, @silene, @herbelin, @JasonGross

### LINKED ISSUE #11091: Parsing/printing inconsistencies and anomalies with notations for partial applications - what to do?
#### Description of the problem

While refactoring `constrextern.ml`, I found the following:
```
Axiom p : forall (a a':nat) {B} (b:B), a=a' /\ b=b.
Notation "## q" := (@p q) (at level 0, q at level 0).
Check ## 3 4 _ true.
(* Printed: "Check ## 3 4 true" *)
Check ## 3 4.
(* anomaly assertion failure in constrextern.ml, since 8.5 *)
```

This raises the question of what to do in the following cases of a notation for a partial application of a constant with implicit arguments:
```
Axiom p : forall (a a':nat) {B} (b:B), a=a' /\ b=b.
Notation "#" := @p (at level 0).
Notation "#'" := p (at level 0).
Notation "## q" := (@p q) (at level 0, q at level 0).
Notation "##' q" := (p q) (at level 0, q at level 0).
Notation P := @p.
Notation P' := p.
Notation Q q := (@p q).
Notation Q' q := (p q).
```
Currently, the (empirical) semantics wrt using the implicit status of arguments to parse the extra arguments is:

|   | parsing | printing | 
|---|---|---|
| `#`  |  no | notation unused |
| `#'`  | yes  | yes |
| `##`  |  no | yes  |
|  `##'` |  no | yes  |
|  `P` |  no |  yes |
|  `P'` |  yes |  yes |
|  `Q` |  yes |  yes |
|  `Q'` |  yes |  yes |

I'm a bit confused about to do. Have heavy users of notations some experience to share? 

A first possibility could be to implement the following table:

|   | parsing | printing | 
|---|---|---|
| `#`  |  no | no |
| `#'`  | yes  | yes |
| `##`  |  no | no  |
|  `##'` |  yes | yes  |
|  `P` |  no |  no |
|  `P'` |  yes |  yes |
|  `Q` |  no |  no |
|  `Q'` |  yes |  yes |

I.e., the notation propagates the implicit argument signature iff there is a no explicit `@` in the right-hand side of the notation.

However, this [comment](https://github.com/coq/coq/issues/4690#issuecomment-337541818) in #4690 (even though #4690 is more generally about the effect of `@g` when `g` is a notation variable) says that Fiat-crypto needs the line `Q`/yes/yes.

One of the issues is that using `@` is the only way to prevent the insertion of maximal implicit arguments (as in `@p 0 0` for `p` as above, or `Notation foo := (@bar _)` in @jasongross example). So, this defeats the idea of using `@` to tell if the resulting notation will inherit the implicit signature of the original constant. An alternative would be to support assigning a specific implicit signature to abbreviations (using `Arguments`) so as to change the default interpretation. Would there be in interest in that?

Added: new proposal which requires two changes in printing and two changes (hopefully corresponding to rare instances) in parsing:

|   | parsing | printing | 
|---|---|---|
| `#`  |  no | no |
| `#'`  | yes  | yes |
| `##`  |  yes | yes  |
|  `##'` |  yes | yes  |
|  `P` |  no |  no |
|  `P'` |  yes |  yes |
|  `Q` |  yes |  yes |
|  `Q'` |  yes |  yes |

#### Side remark: parentheses blocking the interpretation of implicit arguments

Giving `(p 2 2) bool true` is accepted: indeed, the parentheses stop the use of implicit arguments. However, in `(p) 2 2 true`,  the parentheses do not block the implicit arguments. I'd be tempted to make that `(p)` is equivalent to `@p` but that may be quite confusing too, since parentheses are commonly thought as having a purely syntactic role. So, I don't know what to think.

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix / cleanup

Fixes #11091. A newly added commit now fixes #4690 too.

That is, it makes inheritance of implicit arguments and argument scopes consistent across parsing and printing for notations bound to partial applications. In particular it decides to always inherit implicit arguments except when the notation is to an expression of the form `@foo` (without arguments), in which case implicit arguments are deactivated. (Formerly, no proper so global decision was taken on this kind of matter resulting in a combination of inconsistent heuristics.)

The PR also proposes to inherit argument scopes in notations bound to an expression of the form `@foo`. This is a change of semantics, but I don't find intuitive that `@` would act on argument scopes. 

The PR incidentally also fixes an index bug in the algorithm deciding when an implicit argument cannot be reinferred and has to be made explicit.
 
So, to summarize, in addition to several printing adjustements, the semantic changes at parsing, which I expect to be rare in practice, are:
- In `Notation "..." := (foo ...)` and `Notation "..." := (@foo ...)`, implicit arguments are now inherited, as it was the case for `Notation id := (foo ...)` and `Notation id := (@foo ...)`, for both terms and patterns.
- In `Notation "..." := @foo` and `Notation id := @foo`, argument scopes are now kept (but implicit arguments continue to be not propagated), for both terms and patterns.

- [X] Added / updated test-suite
- [X] changelog
- [x] document these rules in the chapters on notations

### CODE REVIEW

**Comment by ejgallego** [comment]
Needs rebase; I wonder if the PR could be split as to ease review, minimize future conflicts, and merge quicker.

File: doc/changelog/03-notations/11120-master+refactoring-application-printing.rst [hunk 1]
```diff
@@ -0,0 +1,16 @@
+- **Fixed:** Inheritance of implicit arguments across notations made
+  uniform in parsing and printing. To the exception of notations of
+  the form ``Notation "..." := @foo`` and ``Notation id := @foo`` which
```
Comment (Zimmi48) [inline doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: You could consider using the Sphinx support for things like:

```suggestion
  the form ``Notation @string := @foo`` and ``Notation @ident__1 := @@ident__2`` which
```

Later on you can also use the `@args` token.

File: doc/changelog/03-notations/11120-master+refactoring-application-printing.rst [hunk 2]
```diff
@@ -0,0 +1,16 @@
+- **Fixed:** Inheritance of implicit arguments across notations made
+  uniform in parsing and printing. To the exception of notations of
```
Comment (Zimmi48) [inline doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: This first sentence sounds a bit convoluted to me. What about:

>Parsing and printing now handle inherited implicit arguments in notations the same.

? I'm not sure it fits exactly what you meant. In any case, we should try to aim for an active form and avoid the word "made".

File: doc/changelog/03-notations/11120-master+refactoring-application-printing.rst [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  the form ``Notation "..." := @foo`` and ``Notation id := @foo`` which
+  inhibit implicit arguments, all notations binding a partially
+  applied constant, as e.g. in ``Notation "..." := (foo args)``,
+  or ``Notation "..." := (@foo args)``, or ``Notation id := (foo args)``, or
+  ``Notation id := (@foo args)``, inherits the remaining implicit arguments
```
Comment (Zimmi48) [inline doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: IINM this is "all notations inherit"

```suggestion
  ``Notation id := (@foo args)``, inherit the remaining implicit arguments
```

File: doc/changelog/03-notations/11120-master+refactoring-application-printing.rst (hunk 1 above)
Comment (herbelin) [reply doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: Strictly speaking the second ident should be a qualid. I'll update.

**Review by herbelin — COMMENTED** [review]
Thanks a lot.

File: doc/changelog/03-notations/11120-master+refactoring-application-printing.rst [hunk 4]
```diff
@@ -0,0 +1,16 @@
+- **Fixed:** Inheritance of implicit arguments across notations made
+  uniform in parsing and printing. To the exception of notations of
+  the form ``Notation @string := @foo`` and ``Notation @ident := @@qualid`` which
```
Comment (Zimmi48) [inline doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: Sorry I was a bit sloppy in my examples, you additionally need to replace the double backticks by simple backticks preceded by `:n:`. E.g.:
```suggestion
  the form :n:`Notation @string := @qualid` and :n:`Notation @ident := @@qualid` which
```

Comment (herbelin) [reply doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: OK, I made a change, please check. In particular, is my ``n:`@@qualid` `` correct?

Comment (Zimmi48) [reply doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: I think so yes. However, what is not is the `@foo` which is interpreted as a token `foo` which does not exist, which breaks the build. If you absolutely want to keep `foo` here, and not use a relevant token like `qualid`, then we need to escape the `@`. I am not sure how to do this, though.

File: doc/changelog/03-notations/11120-master+refactoring-application-printing.rst [hunk 5]
```diff
@@ -0,0 +1,16 @@
+- **Fixed:** Inheritance of implicit arguments across notations made
+  uniform in parsing and printing. To the exception of notations of
+  the form :n:`Notation @string := @@qualid` and :n:`Notation @ident := @@qualid` which
+  inhibit implicit arguments, all notations binding a partially
+  applied constant, as e.g. in :n:`Notation @string := (@qualid @args)`,
+  or :n:`Notation @string := (@@qualid @args)`, or :n:`Notation @ident := (@qualid @args)`, or
+  :n:`Notation @ident := (@@qualid @args)`, inherit the remaining implicit arguments
```
Comment (Zimmi48) [inline doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: I didn't realize before but you can further simplify this list in just two:

```rst
as e.g. in :n:`Notation {| @string | @ident } := (@qualid @args)`
or :n:`Notation {| @string | @ident } := (@@qualid @args)`
```

or even in just one (maybe a bit extreme):

```rst
as e.g. in :n:`Notation {| @string | @ident } := ({? @ } @qualid @args)`
```

**Comment by maximedenes** [comment]
> The PR incidentally also fixes an index bug in the algorithm deciding when an implicit argument cannot be reinferred and has to be made explicit.

How hard would it be to extract this fix? It would be good to integrate it, and possibly backport it.

**Comment by herbelin** [comment]
> How hard would it be to extract this fix? It would be good to integrate it, and possibly backport it.

It would not be too difficult (it is a bit of arithmetic done in 18dedf66b54f4 to be propagated by continuity in the `exprec` subfunction of `Constrextern.explicitize`), but it is a maybe relatively minor bug in `Defensive` mode. Was it reported?

File: doc/changelog/03-notations/11120-master+refactoring-application-printing.rst (hunk 2 above)
Comment (jfehrle) [inline doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]:
```suggestion
  uniform in parsing and printing. With the exception of notations of
```

Comment (herbelin) [reply doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: I made the following proposition: "Parsing and printing consistently handle inheritance of implicit arguments in notations". WDYT?

File: doc/changelog/03-notations/11120-master+refactoring-application-printing.rst (hunk 5 above)
Comment (herbelin) [reply doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: I prefer not to make the reading to tricky.

**Comment by herbelin** [comment]
@ejgallego:

> I wonder if the PR could be split as to ease review, minimize future conflicts, and merge quicker.

The four first commits are large but in principle w/o semantics change. They somehow clutter the PR uselessly and I'm going to submit them separately, hoping it helps.

**Comment by herbelin** [comment]
I rebased. How can I help reviewing for the 19 remaining commits? These are:
- 6 fixes (among them the anomaly in #11091 and #4690)
- 6 commits about unifying the semantics along the line of #11901
- 6 commits about doc/changelog/overlays/test-suite
- 1 small extra commit about restructuring code.

**Comment by ejgallego** [comment]
> I rebased. How can I help reviewing for the 19 remaining commits? These are:

Thanks for the summary @herbelin ; from my part I'm gonna try to process all other notation-based PRs, then address this one.

**Comment by ejgallego** [comment]
I had a look, IMHO this looks mostly ready and it addresses and important problem; I have a few comments about the code I'll do shortly.

File: interp/constrextern.ml [hunk 6]
```diff
@@ -736,8 +736,10 @@ let extern_applied_notation n impl f args =
   if List.is_empty args then
     f.CAst.v
   else
-    let args = adjust_implicit_arguments false (List.length args) 1 args impl in
+  try
```
Comment (ejgallego) [inline interp/constrextern.ml]: This `try ... catch` was missing due to the API of adjust_implicit_arguments? Using an exception here seems overkill and prone to errors maybe.

File: interp/notation_ops.mli [hunk 7]
```diff
@@ -69,12 +69,12 @@ val match_notation_constr : bool -> 'a glob_constr_g -> interpretation ->
 val match_notation_constr_cases_pattern :
   'a cases_pattern_g -> interpretation ->
   (('a cases_pattern_g * extended_subscopes) list * ('a cases_pattern_g list * extended_subscopes) list) *
-    (int * 'a cases_pattern_g list)
+    (bool * int * 'a cases_pattern_g list)
```
Comment (ejgallego) [inline interp/notation_ops.mli]: This is becoming quite hard to read; maybe time to use a record?

**Review by ejgallego — APPROVED** [review]
LGTM, I propose to merge as soon as CI is OK; @herbelin have you opened the overlay's PRs?

I cannot see them.

**Comment by ejgallego** [comment]
> @herbelin have you opened the overlay's PRs?

overlays links still missing

**Comment by ejgallego** [comment]
Ups, closed on error sorry. Refman is broken:
```
Warning, treated as error:
 doc/unreleased.rst:53:'token' reference target not found: args
```

File: doc/changelog/03-notations/11120-master+refactoring-application-printing.rst (hunk 5 above)
Comment (herbelin) [reply doc/changelog/03-notations/11120-master+refactoring-application-printing.rst]: Apparently, sphinx fails on `@args` (`'token' reference target not found`). I replaced by `{+ @arg }` which seems to be the standard way to express what I meant (a non-empty list of arguments).

**Comment by herbelin** [comment]
> have you opened the overlay's PRs?

They were, and, since they were backward-compatible, they were merged at once. The problem is that the overlays were now out of sync. I simply removed the overlay commit.

>  Refman is broken:

I fixed the failure of the changelog.

Finally, I just added a documentation of the rule for inheriting implicit arguments and scopes.

**Comment by ejgallego** [comment]
> Finally, I just added a documentation of the rule for inheriting implicit arguments and scopes.

Great, thanks @herbelin ; will merge once the CI is done.
